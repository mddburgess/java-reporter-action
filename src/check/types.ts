export enum RunCondition {
  disabled,
  autodetect,
  expected,
  required,
}
