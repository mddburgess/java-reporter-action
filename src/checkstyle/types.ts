import { relativePath } from "../common/utils";
import { AnnotationLevel } from "../github/types";
import LintAnnotation from "../results/LintAnnotation";

export default interface CheckstyleReport {
  violations: CheckstyleViolation[];
}

export interface CheckstyleViolation {
  filePath: string;
  line: number;
  column: number;
  rule: string;
  severity: CheckstyleSeverity;
  message: string;
}

export type CheckstyleSeverity = "error" | "warning" | "info";

export const toLintAnnotation = (violation: CheckstyleViolation): LintAnnotation => ({
  path: relativePath(violation.filePath),
  line: violation.line,
  level: getLevel(violation.severity),
  category: getCategory(violation.rule),
  type: getType(violation.rule),
  message: violation.message,
});

const getLevel = (severity: CheckstyleSeverity): AnnotationLevel => {
  switch (severity) {
    case "error":
      return "failure";
    case "warning":
      return "warning";
    case "info":
      return "notice";
    default:
      throw Error("Unexpected severity");
  }
};

const getCategory = (rule: string): string => {
  const idx = rule.lastIndexOf(".");
  return idx === -1 ? "" : rule.slice(0, idx);
};

const getType = (rule: string): string => {
  return rule.split(".").slice(-1)[0];
};
