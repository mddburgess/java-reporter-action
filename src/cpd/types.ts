export default interface CpdReport {
  duplications: CpdDuplication[];
}

export interface CpdDuplication {
  lines: number;
  tokens: number;
  files: CpdFile[];
}

export interface CpdFile {
  path: string;
  startLine: number;
  endLine: number;
  startColumn: number;
  endColumn: number;
}
