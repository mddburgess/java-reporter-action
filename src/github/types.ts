interface CheckRequest {
  status?: CheckStatus;
  started_at?: string;
  conclusion?: CheckConclusion;
  completed_at?: string;
  output?: CheckOutput;
}

type CreateCheckRequest = CheckRequest & { name: string };

type UpdateCheckRequest = CheckRequest & { check_run_id: number };

type CheckStatus = "queued" | "in_progress" | "completed";

type CheckConclusion =
  | "success"
  | "failure"
  | "neutral"
  | "cancelled"
  | "skipped"
  | "timed_out"
  | "action_required";

interface CheckOutput {
  title: string;
  summary: string;
  text?: string;
  annotations?: CheckAnnotation[];
}

interface CheckAnnotation {
  path: string;
  start_line: number;
  end_line: number;
  start_column?: number;
  end_column?: number;
  annotation_level: AnnotationLevel;
  message: string;
  title?: string;
  raw_details?: string;
}

type AnnotationLevel = "notice" | "warning" | "failure";

export {
  CheckRequest,
  CheckConclusion,
  CheckAnnotation,
  CreateCheckRequest,
  UpdateCheckRequest,
};
