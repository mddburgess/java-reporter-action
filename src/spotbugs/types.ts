export default interface SpotbugsReport {
  categories: Map<string, string>;
  bugs: SpotbugsBug[];
}

export interface SpotbugsBug {
  filePath: string;
  startLine: number;
  endLine: number;
  category: string;
  priority: number;
  shortMessage: string;
  longMessage: string;
}
