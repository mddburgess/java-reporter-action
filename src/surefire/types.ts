export default interface SurefireReport {
  name: string;
  tests: number;
  passed: number;
  failures: number;
  errors: number;
  skipped: number;
  testCases: SurefireTestCase[];
}

export interface SurefireTestCase {
  className: string;
  testName: string;
  result?: SurefireTestResult;
  message?: string;
  stackTrace?: string;
}

export type SurefireTestResult = "success" | "failure" | "error" | "skipped";
