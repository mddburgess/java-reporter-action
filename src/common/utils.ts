export function plural(quantity: number, noun: string): string {
  return quantity === 1 ? `${quantity} ${noun}` : `${quantity} ${noun}s`;
}
