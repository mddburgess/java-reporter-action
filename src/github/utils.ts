import { AnnotationLevel, CheckAnnotation } from "./types";

export const compareAnnotations = (a: CheckAnnotation, b: CheckAnnotation): number => {
  if (a.annotation_level !== b.annotation_level) {
    return levelValue(a.annotation_level) - levelValue(b.annotation_level);
  }
  if (a.path !== b.path) {
    return a.path.localeCompare(b.path);
  }
  if (a.start_line !== b.start_line) {
    return a.start_line - b.start_line;
  }
  return 0;
};

const levelValue = (level: AnnotationLevel) => {
  switch (level) {
    case "failure":
      return 1;
    case "warning":
      return 2;
    case "notice":
      return 3;
    default:
      throw Error("Unknown annotation level");
  }
};
