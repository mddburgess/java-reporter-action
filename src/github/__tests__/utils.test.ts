import { CheckAnnotation } from "../types";
import { compareAnnotations } from "../utils";

describe("compareAnnotations()", () => {
  it("sorts annotations by level", () => {
    const annotations: CheckAnnotation[] = [
      {
        path: "",
        start_line: 0,
        end_line: 0,
        annotation_level: "notice",
        message: "",
      },
      {
        path: "",
        start_line: 0,
        end_line: 0,
        annotation_level: "warning",
        message: "",
      },
      {
        path: "",
        start_line: 0,
        end_line: 0,
        annotation_level: "failure",
        message: "",
      },
    ];

    annotations.sort(compareAnnotations);
    expect(annotations).toMatchSnapshot();
  });

  it("sorts annotations with same level by path", () => {
    const annotations: CheckAnnotation[] = [
      {
        path: "path/to/package/Second.java",
        start_line: 0,
        end_line: 0,
        annotation_level: "warning",
        message: "",
      },
      {
        path: "path/to/package/nested/Third.java",
        start_line: 0,
        end_line: 0,
        annotation_level: "warning",
        message: "",
      },
      {
        path: "path/to/package/First.java",
        start_line: 0,
        end_line: 0,
        annotation_level: "warning",
        message: "",
      },
    ];

    annotations.sort(compareAnnotations);
    expect(annotations).toMatchSnapshot();
  });

  it("sorts annotations with same path by start line", () => {
    const annotations: CheckAnnotation[] = [
      {
        path: "path/to/Class.java",
        start_line: 2,
        end_line: 2,
        annotation_level: "warning",
        message: "",
      },
      {
        path: "path/to/Class.java",
        start_line: 3,
        end_line: 3,
        annotation_level: "warning",
        message: "",
      },
      {
        path: "path/to/Class.java",
        start_line: 1,
        end_line: 1,
        annotation_level: "warning",
        message: "",
      },
    ];

    annotations.sort(compareAnnotations);
    expect(annotations).toMatchSnapshot();
  });

  it("handles identical annotations", () => {
    const annotations: CheckAnnotation[] = [
      {
        path: "path/to/Same.java",
        start_line: 1,
        end_line: 1,
        annotation_level: "warning",
        message: "",
      },
      {
        path: "path/to/Different.java",
        start_line: 2,
        end_line: 2,
        annotation_level: "warning",
        message: "",
      },
      {
        path: "path/to/Same.java",
        start_line: 1,
        end_line: 1,
        annotation_level: "warning",
        message: "",
      },
    ];

    annotations.sort(compareAnnotations);
    expect(annotations).toMatchSnapshot();
  });

  it("handles an array of length 1", () => {
    const annotations: CheckAnnotation[] = [
      {
        path: "",
        start_line: 0,
        end_line: 0,
        annotation_level: "notice",
        message: "",
      },
    ];

    annotations.sort(compareAnnotations);
    expect(annotations).toMatchSnapshot();
  });

  it("handles an empty array", () => {
    const annotations: CheckAnnotation[] = [];
    annotations.sort(compareAnnotations);
    expect(annotations).toMatchSnapshot();
  });
});
