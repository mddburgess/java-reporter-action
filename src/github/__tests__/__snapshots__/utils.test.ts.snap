// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compareAnnotations() handles an array of length 1 1`] = `
Array [
  Object {
    "annotation_level": "notice",
    "end_line": 0,
    "message": "",
    "path": "",
    "start_line": 0,
  },
]
`;

exports[`compareAnnotations() handles an empty array 1`] = `Array []`;

exports[`compareAnnotations() handles identical annotations 1`] = `
Array [
  Object {
    "annotation_level": "warning",
    "end_line": 2,
    "message": "",
    "path": "path/to/Different.java",
    "start_line": 2,
  },
  Object {
    "annotation_level": "warning",
    "end_line": 1,
    "message": "",
    "path": "path/to/Same.java",
    "start_line": 1,
  },
  Object {
    "annotation_level": "warning",
    "end_line": 1,
    "message": "",
    "path": "path/to/Same.java",
    "start_line": 1,
  },
]
`;

exports[`compareAnnotations() sorts annotations by level 1`] = `
Array [
  Object {
    "annotation_level": "failure",
    "end_line": 0,
    "message": "",
    "path": "",
    "start_line": 0,
  },
  Object {
    "annotation_level": "warning",
    "end_line": 0,
    "message": "",
    "path": "",
    "start_line": 0,
  },
  Object {
    "annotation_level": "notice",
    "end_line": 0,
    "message": "",
    "path": "",
    "start_line": 0,
  },
]
`;

exports[`compareAnnotations() sorts annotations with same level by path 1`] = `
Array [
  Object {
    "annotation_level": "warning",
    "end_line": 0,
    "message": "",
    "path": "path/to/package/First.java",
    "start_line": 0,
  },
  Object {
    "annotation_level": "warning",
    "end_line": 0,
    "message": "",
    "path": "path/to/package/nested/Third.java",
    "start_line": 0,
  },
  Object {
    "annotation_level": "warning",
    "end_line": 0,
    "message": "",
    "path": "path/to/package/Second.java",
    "start_line": 0,
  },
]
`;

exports[`compareAnnotations() sorts annotations with same path by start line 1`] = `
Array [
  Object {
    "annotation_level": "warning",
    "end_line": 1,
    "message": "",
    "path": "path/to/Class.java",
    "start_line": 1,
  },
  Object {
    "annotation_level": "warning",
    "end_line": 2,
    "message": "",
    "path": "path/to/Class.java",
    "start_line": 2,
  },
  Object {
    "annotation_level": "warning",
    "end_line": 3,
    "message": "",
    "path": "path/to/Class.java",
    "start_line": 3,
  },
]
`;
